src/transport_picoquic.h

```diff
@@ -0,0 +1,159 @@
+
+#pragma once
+
+#include <atomic>
+#include <cassert>
+#include <cstdint>
+#include <vector>
+#include <map>
+#include <mutex>
+#include <queue>
+#include <string>
+#include <thread>
+#include <functional>
+
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+
+#include <transport/transport.h>
+#include <picoquic.h>
+#include <picoquic_config.h>
+#include <picoquic_packet_loop.h>
+
+#include "safe_queue.h"
+
+namespace qtransport {
+
+class PicoQuicTransport : public ITransport
+{
+public:
+  const char* QUICR_ALPN = "quicr-v1";
+
+  struct OutData {
+    std::vector<uint8_t> bytes;
+  };
+
+  struct StreamContext {
+    uint64_t stream_id;
+    TransportContextId context_id;
+    picoquic_cnx_t *cnx;
+    char peer_addr_text[45];
+    uint16_t peer_port;
+    safeQueue<std::vector<uint8_t>> in_data;      /// Pending messages to dequeue
+    safeQueue<OutData> out_data;                  /// Pending messages from enqueue - Only used for datagram
+  };
+
+
+  /*
+   * Exceptions
+   */
+  struct Exception : public std::runtime_error
+  {
+    using std::runtime_error::runtime_error;
+  };
+
+  struct InvalidConfigException : public Exception
+  {
+    using Exception::Exception;
+  };
+
+  struct PicoQuicException : public Exception
+  {
+    using Exception::Exception;
+  };
+
+public:
+  PicoQuicTransport(const TransportRemote& server,
+                    const TransportConfig& tcfg,
+                    TransportDelegate& delegate,
+                    bool isServerMode,
+                    LogHandler& logger);
+
+  virtual ~PicoQuicTransport();
+
+  TransportStatus status() override;
+
+  TransportContextId start() override;
+
+  void close(const TransportContextId& context_id) override;
+  void closeStream(const TransportContextId& context_id,
+                   StreamId stream_id) override;
+
+  StreamId createStream(const TransportContextId& context_id,
+                        bool use_reliable_transport) override;
+
+  TransportError enqueue(const TransportContextId& context_id,
+                         const StreamId & stream_id,
+                         std::vector<uint8_t>&& bytes) override;
+
+
+  std::optional<std::vector<uint8_t>> dequeue(
+    const TransportContextId& context_id,
+    const StreamId & stream_id) override;
+
+  /*
+   * Internal public methods
+   */
+  void setStatus(TransportStatus status);
+
+  StreamContext * getZeroStreamContext(picoquic_cnx_t* cnx);
+
+  StreamContext * createStreamContext(picoquic_cnx_t* cnx,
```

Comment:
Why return raw pointers?

Reply:
It's because of having to pass them to C functions (picoquic).